File: src/test/java/CardTest.java
import org.junit.Assert;
import org.junit.Test;

public class CardTest {

    @Test(expected = IllegalArgumentException.class)
    public void test_whenInputInvalidFormat_shouldThrowException() {
        Card.fromLine("2|green|[2,3,0,0,2]");
    }

    @Test(expected = IllegalArgumentException.class)
    public void test_whenIncludeUnexpectCharacter_shouldThrowException() {
        Card.fromLine("2.1|green|[2,3,0,0,2]");
    }


    @Test(expected = IllegalArgumentException.class)
    public void test_whenIncludeUnexpectSpecialCharacter_shouldThrowException() {
        Card.fromLine("2.|1|green|[2,3,0,0,2]");
    }

    @Test
    public void test_whenInputCharacterFiledGem_shouldCharacterIsLowerOnly() {
        Card subject = Card.fromLine("2|1|green|[2,3,0,0,2]");
        Assert.assertEquals(Gem.GREEN, subject.getResource());
    }

    @Test
    public void test_whenInputValidWBB() {
        Card card = Card.fromLine("2|1|green|[2,3,0,0,2]");

        Assert.assertNotNull(card);
        Assert.assertEquals(2, card.getLevel());
        Assert.assertEquals(1, card.getScore());
        Assert.assertEquals(Gem.GREEN, card.getResource());
        Assert.assertEquals(3, card.getPrice().size());

        Assert.assertEquals(2, card.getPrice().get(Gem.WHITE).longValue());
        Assert.assertEquals(3, card.getPrice().get(Gem.BLUE).longValue());
        Assert.assertEquals(2, card.getPrice().get(Gem.BLACK).longValue());
    }

    @Test
    public void test_whenInputValidRGB() {
        Card card = Card.fromLine("2|2|black|[0,1,4,2,0]");

        Assert.assertNotNull(card);
        Assert.assertEquals(2, card.getLevel());
        Assert.assertEquals(2, card.getScore());
        Assert.assertEquals(Gem.BLACK, card.getResource());
        Assert.assertEquals(3, card.getPrice().size());

        Assert.assertEquals(2, card.getPrice().get(Gem.RED).longValue());
        Assert.assertEquals(4, card.getPrice().get(Gem.GREEN).longValue());
        Assert.assertEquals(1, card.getPrice().get(Gem.BLUE).longValue());
    }




}

File: src/test/kotlin/CardTest.kt
import domain.Card
import domain.Gem
import org.junit.Assert
import org.junit.Test

class CardTest {

    @Test(expected = IllegalArgumentException::class)
    fun `When input invalid format - Should throw exception`() {
        Card.fromLine("2|green|[2,3,0,0,2]")
    }

    @Test(expected = IllegalArgumentException::class)
    fun `When include unexpect character - Should throw exception`() {
        Card.fromLine("2.1|green|[2,3,0,0,2]")
    }

    @Test(expected = IllegalArgumentException::class)
    fun `When include unexpect special character - Should throw exception`() {
        Card.fromLine("2.|1|green|[2,3,0,0,2]")
    }

    @Test
    fun `When input character filed gem - Should character is lower only`() {
        val subject = Card.fromLine("2|1|green|[2,3,0,0,2]")
        Assert.assertEquals(Gem.GREEN, subject.resource)
    }

    @Test
    fun `When input valid - Should be WBB`() {
        val card = Card.fromLine("2|1|green|[2,3,0,0,2]")

        Assert.assertNotNull(card)
        Assert.assertEquals(2, card.level)
        Assert.assertEquals(1, card.score)
        Assert.assertEquals(Gem.GREEN, card.resource)
        Assert.assertEquals(3, card.prices.size)

        Assert.assertEquals(2, card.prices[Gem.WHITE])
        Assert.assertEquals(3, card.prices[Gem.BLUE])
        Assert.assertEquals(2, card.prices[Gem.BLACK])
    }

    @Test
    fun `When input valid - Should be RGB`() {
        val card = Card.fromLine("2|2|black|[0,1,4,2,0]")

        Assert.assertNotNull(card)
        Assert.assertEquals(2, card.level)
        Assert.assertEquals(2, card.score)
        Assert.assertEquals(Gem.BLACK, card.resource)
        Assert.assertEquals(3, card.prices.size)

        Assert.assertEquals(2, card.prices[Gem.RED])
        Assert.assertEquals(4, card.prices[Gem.GREEN])
        Assert.assertEquals(1, card.prices[Gem.BLUE])
    }


}
